## Spotting missing values

We analyzed the San Francisco building permits data set in terms of missing values. We found the missing values, show how many there are and also suggest how to remove them.

*Advanced: We also share our opinion whether missing values should be removed in this case or not and justify our answer.*





**Start of the Project**

In [1]:
import pandas as pd

# Load the data
url = 'https://drive.google.com/file/d/1G4BANCCnsCd9hayjdbNq3fvRgjnhGblz/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
data = pd.read_csv(path)

# Display the first 3 rows of the dataset
data.head(3)

/usr/lib/python3/dist-packages/IPython/core/interactiveshell.py:3062: DtypeWarning: Columns (22,32) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   

  Street Name Street Suffix  ...  Existing Construction Type  \
0       Ellis            St  ...                         3.0   
1       Geary            St  ...                         3.0   
2     Pacific            Av  ...                         1.0   

  Existing Construction Type Description Proposed Construction Type  \
0                          constr type 3                        NaN   
1                          constr type 3                        NaN   
2                          constr type 1                        1.0   

  Proposed Construction Type Description Site Permit Supervisor District  \
0                                    NaN         NaN                 3.0   
1                                    NaN         NaN                 3.0   
2                          constr type 1         NaN                 3.0   

  Neighborhoods - Analysis Boundaries  Zipcode  \
0                          Tenderloin  94102.0   
1                          Tenderloin  94102.0   
2                        Russian Hill  94109.0   

                                    Location      Record ID  
0  (37.785719256680785, -122.40852313194863)  1380611233945  
1   (37.78733980600732, -122.41063199757738)  1420164406718  
2    (37.7946573324287, -122.42232562979227)  1424856504716  

[3 rows x 43 columns]

In [2]:
# Check the shape of the dataset 
data.shape

(198900, 43)

From the previous result, there are 198900 observations (rows) and 43 variables (columns). Let's retrieve more info from the dataset.

In [3]:
# Check data types
data.dtypes

Permit Number                              object
Permit Type                                 int64
Permit Type Definition                     object
Permit Creation Date                       object
Block                                      object
Lot                                        object
Street Number                               int64
Street Number Suffix                       object
Street Name                                object
Street Suffix                              object
Unit                                      float64
Unit Suffix                                object
Description                                object
Current Status                             object
Current Status Date                        object
Filed Date                                 object
Issued Date                                object
Completed Date                             object
First Construction Document Date           object
Structural Notification                    object


Variable types vary from numeric (*int64*, *float64*) to *object* which is a generic type and may need to be addressed. 

In [4]:
# Create an index for columns which data type is 'object'
obj_type = data.columns[data.dtypes == 'object']

# Display columns which data type is 'object'
data[obj_type]

Permit Number            Permit Type Definition Permit Creation Date  \
0       201505065519                      sign - erect           05/06/2015   
1       201604195146                      sign - erect           04/19/2016   
2       201605278609  additions alterations or repairs           05/27/2016   
3       201611072166            otc alterations permit           11/07/2016   
4       201611283529                       demolitions           11/28/2016   
...              ...                               ...                  ...   
198895       M862628            otc alterations permit           12/05/2017   
198896  201712055595            otc alterations permit           12/05/2017   
198897       M863507            otc alterations permit           12/06/2017   
198898       M863747            otc alterations permit           12/06/2017   
198899       M864287            otc alterations permit           12/07/2017   

       Block   Lot Street Number Suffix Street Name Street Suffix Unit Suffix  \
0       0326   023                  NaN       Ellis            St         NaN   
1       0306   007                  NaN       Geary            St         NaN   
2       0595   203                  NaN     Pacific            Av         NaN   
3       0156   011                  NaN     Pacific            Av         NaN   
4       0342   001                  NaN      Market            St         NaN   
...      ...   ...                  ...         ...           ...         ...   
198895  0113  017A                  NaN  Montgomery            St         NaN   
198896  0271   014                  NaN        Bush            St         NaN   
198897  4318   019                  NaN     Indiana            St         NaN   
198898  0298   029                  NaN      Sutter            St         NaN   
198899  0160   006                  NaN     Pacific            Av         NaN   

                                              Description  ...  \
0       ground fl facade: to erect illuminated, electr...  ...   
1                 remove (e) awning and associated signs.  ...   
2                         installation of separating wall  ...   
3                repair dryrot & stucco at front of bldg.  ...   
4       demolish retail/office/commercial 3-story buil...  ...   
...                                                   ...  ...   
198895                                       street space  ...   
198896                fire alarm upgrade ref 201704123852  ...   
198897                                       street space  ...   
198898                                street space permit  ...   
198899                                street space permit  ...   

       Fire Only Permit Permit Expiration Date         Existing Use  \
0                   NaN             11/03/2016  tourist hotel/motel   
1                   NaN             12/03/2017  tourist hotel/motel   
2                   NaN                    NaN         retail sales   
3                   NaN             07/13/2018    1 family dwelling   
4                   NaN             12/01/2018         retail sales   
...                 ...                    ...                  ...   
198895              NaN                    NaN                  NaN   
198896                Y             04/06/2018           apartments   
198897              NaN                    NaN                  NaN   
198898              NaN                    NaN                  NaN   
198899              NaN                    NaN                  NaN   

             Proposed Use TIDF Compliance  \
0                     NaN             NaN   
1                     NaN             NaN   
2            retail sales             NaN   
3       1 family dwelling             NaN   
4                     NaN             NaN   
...                   ...             ...   
198895                NaN             NaN   
198896         apartments             NaN   
198897                NaN         

As per the last result, most of the variables that type is *object* have strings as their values or a mix of numbers and characters. However, some of them are dates. Since this is not a time series dataset, the dates here function as a characteristic of the observation.

Now let's check on missing values.

In [5]:
# Create an index for columns which have missing values
has_nan = data.columns[data.isna().any()]

# Display columns with missing values
data[has_nan]

Street Number Suffix Street Suffix  Unit Unit Suffix  \
0                       NaN            St   NaN         NaN   
1                       NaN            St   0.0         NaN   
2                       NaN            Av   NaN         NaN   
3                       NaN            Av   0.0         NaN   
4                       NaN            St   NaN         NaN   
...                     ...           ...   ...         ...   
198895                  NaN            St   NaN         NaN   
198896                  NaN            St   NaN         NaN   
198897                  NaN            St   NaN         NaN   
198898                  NaN            St   NaN         NaN   
198899                  NaN            Av   NaN         NaN   

                                              Description Issued Date  \
0       ground fl facade: to erect illuminated, electr...  11/09/2015   
1                 remove (e) awning and associated signs.  08/03/2017   
2                         installation of separating wall         NaN   
3                repair dryrot & stucco at front of bldg.  07/18/2017   
4       demolish retail/office/commercial 3-story buil...  12/01/2017   
...                                                   ...         ...   
198895                                       street space  12/05/2017   
198896                fire alarm upgrade ref 201704123852  12/06/2017   
198897                                       street space  12/06/2017   
198898                                street space permit  12/06/2017   
198899                                street space permit  12/07/2017   

       Completed Date First Construction Document Date  \
0                 NaN                       11/09/2015   
1                 NaN                       08/03/2017   
2                 NaN                              NaN   
3          07/24/2017                       07/18/2017   
4                 NaN                       11/20/2017   
...               ...                              ...   
198895            NaN                       12/05/2017   
198896            NaN                       12/06/2017   
198897            NaN                       12/06/2017   
198898            NaN                       12/06/2017   
198899            NaN                       12/07/2017   

       Structural Notification  Number of Existing Stories  ...  \
0                          NaN                         6.0  ...   
1                          NaN                         7.0  ...   
2                          NaN                         6.0  ...   
3                          NaN                         2.0  ...   
4                          NaN                         3.0  ...   
...                        ...                         ...  ...   
198895                     NaN                         NaN  ...   
198896                     NaN                         4.0  ...   
198897                     NaN                         NaN  ...   
198898                     NaN                         NaN  ...   
198899                     NaN                         NaN  ...   

        TIDF Compliance Existing Construction Type  \
0                   NaN                        3.0   
1                   NaN                        3.0   
2                   NaN                        1.0   
3                   NaN                        5.0   
4                   NaN                        3.0   
...                 ...                        ...   
198895              NaN                        NaN   
198896              NaN                        5.0   
198897              NaN                        NaN   
198898              NaN                        NaN   
198899              NaN                        NaN   

       Existing Construction Type Description Proposed Construction Type  \
0                               constr type 3                        NaN   
1                               constr type 3                        N

From the total of 43 columns, 31 have missing values.

In [6]:
# Display the number of missing values
data[has_nan].isna().sum()

Street Number Suffix                      196684
Street Suffix                               2768
Unit                                      169421
Unit Suffix                               196939
Description                                  290
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Structural Notification                   191978
Number of Existing Stories                 42784
Number of Proposed Stories                 42868
Voluntary Soft-Story Retrofit             198865
Fire Only Permit                          180073
Permit Expiration Date                     51880
Estimated Cost                             38066
Revised Cost                                6066
Existing Use                               41114
Existing Units                             51538
Proposed Use                               42439
Proposed Units                             50911
Plansets            

As noticed from the previous results, the number of missing values varies substantially depending on which variable we are considering. For instance, the *Description* column has only 290 missing values whereas *TIDF Compliance* has 198898 (almost the total of rows). At this point, we can take different approaches.

### 1. Delete Rows/Collumns with Missing Values

Remove the observations (rows) with missing values. This approach seems to be the simplest (and fastest) solution for the issue of missing values. It also keeps all the variables.

In [7]:
# Make a copy of the dataset
data1 = data.copy()

# Drop the rows with missing values
data1 = data1.dropna(axis=0,how='any')

# Check the result
data1

Empty DataFrame
Columns: [Permit Number, Permit Type, Permit Type Definition, Permit Creation Date, Block, Lot, Street Number, Street Number Suffix, Street Name, Street Suffix, Unit, Unit Suffix, Description, Current Status, Current Status Date, Filed Date, Issued Date, Completed Date, First Construction Document Date, Structural Notification, Number of Existing Stories, Number of Proposed Stories, Voluntary Soft-Story Retrofit, Fire Only Permit, Permit Expiration Date, Estimated Cost, Revised Cost, Existing Use, Existing Units, Proposed Use, Proposed Units, Plansets, TIDF Compliance, Existing Construction Type, Existing Construction Type Description, Proposed Construction Type, Proposed Construction Type Description, Site Permit, Supervisor District, Neighborhoods - Analysis Boundaries, Zipcode, Location, Record ID]
Index: []

[0 rows x 43 columns]

The last step was not feasible since every observation (row) has at least one missing value and we ended up with an empty dataset. Now, let's remove the variables (columns) with missing values. 

In [8]:
# Make a copy of the dataset
data2 = data.copy()

# Drop the columns with missing values
data2 = data2.dropna(axis=1,how='any')

# Check the result
data2

Permit Number  Permit Type            Permit Type Definition  \
0       201505065519            4                      sign - erect   
1       201604195146            4                      sign - erect   
2       201605278609            3  additions alterations or repairs   
3       201611072166            8            otc alterations permit   
4       201611283529            6                       demolitions   
...              ...          ...                               ...   
198895       M862628            8            otc alterations permit   
198896  201712055595            8            otc alterations permit   
198897       M863507            8            otc alterations permit   
198898       M863747            8            otc alterations permit   
198899       M864287            8            otc alterations permit   

       Permit Creation Date Block   Lot  Street Number Street Name  \
0                05/06/2015  0326   023            140       Ellis   
1                04/19/2016  0306   007            440       Geary   
2                05/27/2016  0595   203           1647     Pacific   
3                11/07/2016  0156   011           1230     Pacific   
4                11/28/2016  0342   001            950      Market   
...                     ...   ...   ...            ...         ...   
198895           12/05/2017  0113  017A           1228  Montgomery   
198896           12/05/2017  0271   014            580        Bush   
198897           12/06/2017  4318   019           1568     Indiana   
198898           12/06/2017  0298   029            795      Sutter   
198899           12/07/2017  0160   006            838     Pacific   

       Current Status Current Status Date  Filed Date      Record ID  
0             expired          12/21/2017  05/06/2015  1380611233945  
1              issued          08/03/2017  04/19/2016  1420164406718  
2           withdrawn          09/26/2017  05/27/2016  1424856504716  
3            complete          07/24/2017  11/07/2016  1443574295566  
4              issued          12/01/2017  11/28/2016   144548169992  
...               ...                 ...         ...            ...  
198895         issued          12/05/2017  12/05/2017  1489337276729  
198896         issued          12/06/2017  12/05/2017  1489462354993  
198897         issued          12/06/2017  12/06/2017  1489539379952  
198898         issued          12/06/2017  12/06/2017  1489608233656  
198899         issued          12/07/2017  12/07/2017  1489796283803  

[198900 rows x 12 columns]

From the total of 43 variables, 12 were left. This may be a feasible solution depending on the final goal. Let's check another approach though.

### 2. Impute missing values with Mean/Median/Mode

Here we will try to impute values for the missing ones. For that, we need to perform a detailed inspection. To speed up the process we will use *Pandas Profiling*.

In [ ]:
!pip install -U pandas-profiling[notebook]

In [12]:
from pandas_profiling import ProfileReport

# Create a profile from the columns with missing values
profile = ProfileReport(data[has_nan], title='Data Profile Report', minimal=True)

# Save data profile report
profile.to_file("data_report.html")

/usr/local/lib/python3.8/dist-packages/pandas/core/frame.py:4300: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


Summarize dataset:   0%|          | 0/39 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

From the dataset [profile](https://rdamatta.github.io/data-quality/data_report.html), some variables like *Structural Notification*, *Voluntary Soft-Story Retrofit*, *Fire Only Permit*, and *Site Permit* can be considered as **boolean** so we can impute *false* (or "N" in this case) for the missing values.

In [ ]:
# Create an index for boolean variables
bol = ['Site Permit','Fire Only Permit','Structural Notification','Voluntary Soft-Story Retrofit']

# Impute "N" for the missing values
data[bol] = data[bol].fillna("N")

Now let's check on **categorical** variables. Starting with *Street Number Suffix*. From the total of 198900 observations, 196684 is missing. The most frequent category is "A" with 1501 occurrences (only 0.75% of total). So, it is a good idea to create a new category like "Z", "absent" or "nil". 

In [ ]:
# Impute "nil" for the missing values
data['Street Number Suffix'] = data['Street Number Suffix'].fillna("nil")

The same reasoning can be applied to other categorical variables.

In [ ]:
# Impute "nil" for the missing values
data['Unit Suffix'] = data['Unit Suffix'].fillna("nil")

Now let's check *Street Suffix* variable. This one is a different situation: the most frequent category is "St" with 138358 occurrences which represent 69.6% of the total number of observations (rows). So, it's reasonable to replace all missing values with it.

The same reasoning can be applied to other categorical variables.

In [ ]:
# Impute the mode for the missing values
data['Street Suffix'] = data['Street Suffix'].fillna("St")
data['Description'] = data['Description'].fillna("street space")
data['Existing Use'] = data['Existing Use'].fillna("1 family dwelling")
data['Proposed Use'] = data['Proposed Use'].fillna("1 family dwelling ")
data['Existing Construction Type'] = data['Existing Construction Type'].fillna(5.0)
data['Existing Construction Type Description'] = data['Existing Construction Type Description'].fillna("wood frame (5)")
data['Proposed Construction Type'] = data['Proposed Construction Type'].fillna(5.0)
data['Proposed Construction Type Description'] = data['Proposed Construction Type Description'].fillna("wood frame (5)")
data['Neighborhoods - Analysis Boundaries'] = data['Neighborhoods - Analysis Boundaries'].fillna("Financial District/South Beach")

There is one last categorical variable which is *TIDF Compliance*. There are only 2 non-missing values which make it unfeasible to perform any imputation. We'll simply drop it.

In [ ]:
# Drop 'TIDF Compliance' columns
data = data.drop(['TIDF Compliance'], axis=1)

Now let's deal with the **numeric** variables. For the *Unit* variable, the distribution is quite skewed to the right. So, it's a good idea to use the *median* to replace the missing values.

In [ ]:
# Impute the median for the missing values
data['Unit'] = data['Unit'].fillna(data['Unit'].median())

The same reasoning can be used for other right-skewed numeric variables.

In [ ]:
# Impute the median for the missing values
data['Number of Proposed Stories'] = data['Number of Proposed Stories'].fillna(data['Number of Proposed Stories'].median())
data['Number of Existing Stories'] = data['Number of Existing Stories'].fillna(data['Number of Existing Stories'].median())
data['Existing Units'] = data['Existing Units'].fillna(data['Existing Units'].median())
data['Proposed Units'] = data['Proposed Units'].fillna(data['Proposed Units'].median())
data['Plansets'] = data['Plansets'].fillna(data['Plansets'].median())


After checking the distribution of the *Estimated Cost* and *Revised Cost* variables, the *mean* is adequate as imputed value for the missing ones.

In [ ]:
# Impute the mean for the missing values
data['Estimated Cost'] = data['Estimated Cost'].fillna(data['Estimated Cost'].mean())
data['Revised Cost'] = data['Revised Cost'].fillna(data['Revised Cost'].mean())

The *Supervisor District* variable has a peculiar distribution. In this case, we'll replace the missing values with the *mode*.

In [ ]:
# Impute the mode for the missing values
data['Supervisor District'] = data['Supervisor District'].fillna(data['Supervisor District'].mode()[0])

The last numeric variables are *Zipcode* and *Location*. There are 1700 missing values for *Location* and 1740 for *Zipcode*. Let's check if these variables are linked with each other.

In [ ]:
# Display the first five rows where Zipcode is missing
data[data['Zipcode'].isnull()].head()

Permit Number  Permit Type  ... Location      Record ID
452       M866367            8  ...      NaN  1490471235302
464       M845627            8  ...      NaN  1490955510104
577       M364427            8  ...      NaN   129220681682
731  201301047237            8  ...      NaN  1292365379064
864  201301077331            8  ...      NaN  1292500249105

[5 rows x 42 columns]

It seems that when *Location* is missing *Zipcode* is also missing. Let's just drop those rows with this missing information.

In [ ]:
# Keep rows which Zipcode is not NaN
data = data[data['Zipcode'].notnull()]

# Keep rows which Location is not NaN
data = data[data['Location'].notnull()]

Finally, let's deal with date-type variables. First, let's check how many missing values are there.

In [ ]:
# Create an index for date variables
date_cols = ['Permit Creation Date','Current Status Date','Filed Date','Issued Date','Completed Date','First Construction Document Date','Permit Expiration Date']

# Count the number of missing values
data[date_cols].isna().sum()

Permit Creation Date                     0
Current Status Date                      0
Filed Date                               0
Issued Date                          14880
Completed Date                      100737
First Construction Document Date     14886
Permit Expiration Date               51222
dtype: int64

Let's check if there any relation between *Current Status* and *Completed Date*.

In [ ]:
# Display the number of observations with Current Status different from 'complete'
data[data['Current Status'] != 'complete'].shape[0]

100851

Yes, there is. If the *Current Status* is different from *complete* there is no *Completed Date*. Depending on the final goal, we may or may not drop those observations with the *Completed Date* missing since they correspond to 51% of the current dataset.

The same occurs with the missing values on *Permit Expiration Date* if the *Current Status* has values like *withdrawn*, *canceled*, *disapproved*, *incomplete*, and so on.

Now, let's count the number of missing values for all variables.

In [ ]:
# Display the number of missing values for ALL variables
data.isna().sum()

Permit Number                                  0
Permit Type                                    0
Permit Type Definition                         0
Permit Creation Date                           0
Block                                          0
Lot                                            0
Street Number                                  0
Street Number Suffix                           0
Street Name                                    0
Street Suffix                                  0
Unit                                           0
Unit Suffix                                    0
Description                                    0
Current Status                                 0
Current Status Date                            0
Filed Date                                     0
Issued Date                                14880
Completed Date                            100737
First Construction Document Date           14886
Structural Notification                        0
Number of Existing S

In [ ]:
# Check the shape of the dataset 
data.shape

(197184, 42)

From the original 198900 observations (rows) we ended up with 197184 which is not a big loss of information.

Worth of note is that there is another approach to impute missing values like the prediction of missing values of one variable according to present values in another variable(s). It tries to find a correlation between variables and build a model upon it.

For the sake of experimenting, let's try this approach.

### 3. Imputation using Machine Learning

Here we'll perform the imputation of categorical columns using *DataWig* library. *DataWig* is a framework for learning models to impute missing values in tables.

We will impute values in a specific column only (called *output_column*) using values in other columns (called *input_columns*). For example, we would like to impute values of *Permit Expiration Date* using the *Issued Date* information.

In [ ]:
!pip install datawig

In [ ]:
import datawig

# Split dataset
df_train, df_test = datawig.utils.random_split(data)

# Initialize a SimpleImputer model
imputer = datawig.SimpleImputer(
    input_columns=['Issued Date'], # column containing information about the column we want to impute
    output_column= 'Permit Expiration Date', # the column we like to impute values for
    output_path = 'imputer_model' # stores model data and metrics
    )

# Fit an imputer model on the train data
imputer.fit(train_df=df_train, num_epochs=10)
 
# Impute missing values and return original dataframe with predictions
imputed = imputer.predict(data)

2021-03-18 10:59:08,339 [INFO]  
========== start: fit model
2021-03-18 10:59:08,344 [WARNING]  Already bound, ignoring bind()
2021-03-18 10:59:42,840 [INFO]  Epoch[0] Batch [0-396]	Speed: 184.28 samples/sec	cross-entropy=2.949706	Permit Expiration Date-accuracy=0.688130
2021-03-18 11:00:21,105 [INFO]  Epoch[0] Train-cross-entropy=2.097556
2021-03-18 11:00:21,106 [INFO]  Epoch[0] Train-Permit Expiration Date-accuracy=0.781527
2021-03-18 11:00:21,114 [INFO]  Epoch[0] Time cost=72.761
2021-03-18 11:00:21,179 [INFO]  Saved checkpoint to "imputer_model/model-0000.params"
2021-03-18 11:00:23,434 [INFO]  Epoch[0] Validation-cross-entropy=0.872676
2021-03-18 11:00:23,444 [INFO]  Epoch[0] Validation-Permit Expiration Date-accuracy=0.897529
2021-03-18 11:01:02,005 [INFO]  Epoch[1] Batch [0-396]	Speed: 164.80 samples/sec	cross-entropy=0.675320	Permit Expiration Date-accuracy=0.910107
2021-03-18 11:01:34,414 [INFO]  Epoch[1] Train-cross-entropy=0.569948
2021-03-18 11:01:34,416 [INFO]  Epoch[1] Tr

From the previous result, our model had quite high accuracy using the validation dataset (about 95% on average). Below we can check that some predictions using the test dataset had achieved a high probability (last column) of getting the right *Permit Expiration Date*. Others not.

In [ ]:
# Display the first 5 rows of predictions
imputed.head()

Permit Number  ...  Permit Expiration Date_imputed_proba
21496   201308134203  ...                              0.516017
43569        M475887  ...                              0.965528
29402   201310290527  ...                              0.998634
148771  201702018433  ...                              0.014284
168315       M883647  ...                              0.999868

[5 rows x 44 columns]

### Conclusion

In this task, we explored 3 approaches to deal with missing values on a dataset composed of numeric, categorical, and date/time values.

The **first approach** has the advantage of being straightforward and easy to deploy but it has the loss of substantial information if the percentage of missing values is excessive in comparison to the dataset as a drawback.

The **second approach** is more time-consuming but it prevents excessive data loss and works well with small datasets. The cons are: it can cause data leakage, and it does not factor in the covariance between variables.

Finally, the **third approach** gives a better result than earlier methods since it takes into account the covariance between entries of a variable we want to deal with its missing values and another variable(s) in the dataset. However, predictions are considered only as a proxy for the true (missing) values.